README.md
termco   
============


[![Project Status: Active - The project has reached a stable, usable
state and is being actively
developed.](http://www.repostatus.org/badges/0.1.0/active.svg)](http://www.repostatus.org/#active)
[![Build
Status](https://travis-ci.org/trinker/termco.svg?branch=master)](https://travis-ci.org/trinker/termco)
[![Coverage
Status](https://coveralls.io/repos/trinker/termco/badge.svg?branch=master)](https://coveralls.io/r/trinker/termco?branch=master)
[![DOI](https://zenodo.org/badge/5398/trinker/termco.svg)](https://zenodo.org/badge/latestdoi/5398/trinker/termco)

![](tools/termco_logo/r_termco.png)

**termco** is a suite of functions used to count and find terms and
substrings in strings. The tools can be used to build an expert rules,
regular expression based text classification model. The package wraps
the [**data.table**](https://cran.r-project.org/package=data.table) and
[**stringi**](https://cran.r-project.org/package=stringi) packages to
create fast data frame counts of regular expression terms and
substrings.


Table of Contents
============

-   [Functions](#functions)
-   [Installation](#installation)
-   [Contact](#contact)
-   [Examples](#examples)
    -   [Load the Tools/Data](#load-the-toolsdata)
    -   [Build Counts Dataframe](#build-counts-dataframe)
    -   [Printing](#printing)
    -   [Plotting](#plotting)
    -   [Ngram Collocations](#ngram-collocations)
        -   [Collocation Plotting](#collocation-plotting)
    -   [Converting to Document Term Matrix](#converting-to-document-term-matrix)
-   [Building an Expert Rules, Regex Classifier Model](#building-an-expert-rules-regex-classifier-model)
    -   [Load the Tools/Data](#load-the-toolsdata-1)
    -   [Splitting Data](#splitting-data)
    -   [Understanding Term Use](#understanding-term-use)
        -   [View Most Used Words](#view-most-used-words)
        -   [View Most Used Words in Context](#view-most-used-words-in-context)
        -   [View Important Words](#view-important-words)
    -   [Building the Model](#building-the-model)
    -   [Testing the Model](#testing-the-model)
    -   [Improving the Model](#improving-the-model)
        -   [Improving Coverage](#improving-coverage)
        -   [Improving Discrimination](#improving-discrimination)
    -   [Categorizing/Tagging](#categorizingtagging)
    -   [Evaluation: Accuracy](#evaluation-accuracy)
        -   [Pre Coded Data](#pre-coded-data)
        -   [Post Coding Data](#post-coding-data)

Functions
============


The main function of **termco** is `term_count`. It is used to extract
regex term counts by grouping variable(s) as well as to generate
classification models.

Most of the functions *count*, *search*, *plot* terms, and *covert*
between output types, while a few remaining functions are used to train,
test and interpret *model*s. Additionally, the `probe_` family of
function generate lists of function calls or plots for given search
terms. The table below describes the functions, category of use, and
their description:

<table>
<colgroup>
<col width="31%" />
<col width="15%" />
<col width="53%" />
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Use Category</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>term_count</code></td>
<td>count</td>
<td>Count regex term occurrence; modeling</td>
</tr>
<tr class="even">
<td><code>token_count</code></td>
<td>count</td>
<td>Count fixed token occurrence; modeling</td>
</tr>
<tr class="odd">
<td><code>frequent_terms</code>/<code>all_words</code></td>
<td>count</td>
<td>Frequent terms</td>
</tr>
<tr class="even">
<td><code>important_terms</code></td>
<td>count</td>
<td>Important terms</td>
</tr>
<tr class="odd">
<td><code>hierarchical_coverage_term</code></td>
<td>count</td>
<td>Unique coverage of a text vector by terms</td>
</tr>
<tr class="even">
<td><code>hierarchical_coverage_regex</code></td>
<td>count</td>
<td>Unique coverage of a text vector by regex</td>
</tr>
<tr class="odd">
<td><code>frequent_ngrams</code></td>
<td>count</td>
<td>Weighted frequent ngram (2 &amp; 3) collocations</td>
</tr>
<tr class="even">
<td><code>word_count</code></td>
<td>count</td>
<td>Count words</td>
</tr>
<tr class="odd">
<td><code>term_before</code>/<code>term_after</code></td>
<td>count</td>
<td>Frequency of words before/after a regex term</td>
</tr>
<tr class="even">
<td><code>term_first</code></td>
<td>count</td>
<td>Frequency of words at the begining of strings</td>
</tr>
<tr class="odd">
<td><code>colo</code></td>
<td>search</td>
<td>Regex output to find term collocations</td>
</tr>
<tr class="even">
<td><code>search_term</code></td>
<td>search</td>
<td>Search for regex terms</td>
</tr>
<tr class="odd">
<td><code>match_word</code></td>
<td>search</td>
<td>Extract words from a text matching a regular expression</td>
</tr>
<tr class="even">
<td><code>search_term_collocations</code></td>
<td>search</td>
<td>Wrapper for <code>search_term</code> + <code>frequent_terms</code></td>
</tr>
<tr class="odd">
<td><code>classification_project</code></td>
<td>modeling</td>
<td>Make a classification modeling project template</td>
</tr>
<tr class="even">
<td><code>classification_template</code></td>
<td>modeling</td>
<td>Make a classification analysis script template</td>
</tr>
<tr class="odd">
<td><code>as_dtm</code>/<code>as_tdm</code></td>
<td>modeling</td>
<td>Coerce <code>term_count</code> object into <code>tm::DocumentTermMatrix</code>/<code>tm::TermDocumentMatrix</code></td>
</tr>
<tr class="even">
<td><code>split_data</code></td>
<td>modeling</td>
<td>Split data into <code>train</code> &amp; <code>test</code> sets</td>
</tr>
<tr class="odd">
<td><code>evaluate</code></td>
<td>modeling</td>
<td>Check accuracy of model against human coder</td>
</tr>
<tr class="even">
<td><code>classify</code></td>
<td>modeling</td>
<td>Assign n tags to text from a model</td>
</tr>
<tr class="odd">
<td><code>get_text</code></td>
<td>modeling</td>
<td>Get the original text for model tags</td>
</tr>
<tr class="even">
<td><code>coverage</code></td>
<td>modeling</td>
<td>Coverage for <code>term_count</code> or <code>search_term</code> object</td>
</tr>
<tr class="odd">
<td><code>uncovered</code>/<code>get_uncovered</code></td>
<td>modeling</td>
<td>Get the uncovered text from a model</td>
</tr>
<tr class="even">
<td><code>mutate_counts</code></td>
<td>modeling</td>
<td>Apply normalizing function to term count columns</td>
</tr>
<tr class="odd">
<td><code>select_counts</code></td>
<td>modeling</td>
<td>Select columns without stripping count classes</td>
</tr>
<tr class="even">
<td><code>tag_co_occurrence</code></td>
<td>modeling</td>
<td>Explore co-occurrence of tags from a model</td>
</tr>
<tr class="odd">
<td><code>validate_model</code>/<code>assign_validation_task</code></td>
<td>modeling</td>
<td>Human validation of a <code>term_count</code> model</td>
</tr>
<tr class="even">
<td><code>read_term_list</code></td>
<td>read/write</td>
<td>Read a term list from an external file</td>
</tr>
<tr class="odd">
<td><code>write_term_list</code></td>
<td>read/write</td>
<td>Write a term list to an external file</td>
</tr>
<tr class="even">
<td><code>term_list_template</code></td>
<td>read/write</td>
<td>Write a term list template to an external file</td>
</tr>
<tr class="odd">
<td><code>as_count</code></td>
<td>convert</td>
<td>Strip pretty printing from <code>term_count</code> object</td>
</tr>
<tr class="even">
<td><code>as_terms</code></td>
<td>convert</td>
<td>Convert a count matrix to list of term vectors</td>
</tr>
<tr class="odd">
<td><code>as_term_list</code></td>
<td>convert</td>
<td>Convert a vector of terms into a named term list</td>
</tr>
<tr class="even">
<td><code>weight</code></td>
<td>convert</td>
<td>Weight a <code>term_count</code> object proportion/percent</td>
</tr>
<tr class="odd">
<td><code>plot_ca</code></td>
<td>plot</td>
<td>Plot <code>term_count</code> object as 3-D correspondence analysis map</td>
</tr>
<tr class="even">
<td><code>plot_counts</code></td>
<td>plot</td>
<td>Horizontal bar plot of group counts</td>
</tr>
<tr class="odd">
<td><code>plot_freq</code></td>
<td>plot</td>
<td>Vertical bar plot of frequencies of counts</td>
</tr>
<tr class="even">
<td><code>plot_cum_percent</code></td>
<td>plot</td>
<td>Plot <code>frequent_terms</code> object as cumulative percent</td>
</tr>
<tr class="odd">
<td><code>probe_list</code></td>
<td>probe</td>
<td>Generate list of <code>search_term</code> function calls</td>
</tr>
<tr class="even">
<td><code>probe_colo_list</code></td>
<td>probe</td>
<td>Generate list of <code>search_term_collocations</code> function calls</td>
</tr>
<tr class="odd">
<td><code>probe_colo_plot_list</code></td>
<td>probe</td>
<td>Generate list of <code>search_term_collocationss</code> + <code>plot</code> function calls</td>
</tr>
<tr class="even">
<td><code>probe_colo_plot</code></td>
<td>probe</td>
<td>Plot <code>probe_colo_plot_list</code> directly</td>
</tr>
</tbody>
</table>

Installation
============

To download the development version of **termco**:

Download the [zip
ball](https://github.com/trinker/termco/zipball/master) or [tar
ball](https://github.com/trinker/termco/tarball/master), decompress and
run `R CMD INSTALL` on it, or use the **pacman** package to install the
development version:

    if (!require("pacman")) install.packages("pacman")
    pacman::p_load_gh(
        "trinker/gofastr",
        "trinker/termco"
    )

Contact
=======

You are welcome to:    
- submit suggestions and bug-reports at: <https://github.com/trinker/termco/issues>    
- send a pull request on: <https://github.com/trinker/termco/>    
- compose a friendly e-mail to: <tyler.rinker@gmail.com>    


Examples
========

The following examples demonstrate some of the functionality of
**termco**.

Load the Tools/Data
-------------------

    if (!require("pacman")) install.packages("pacman")
    pacman::p_load(dplyr, ggplot2, termco)

    data(presidential_debates_2012)

Build Counts Dataframe
----------------------

    discoure_markers <- list(
        response_cries = c("\\boh", "\\bah", "aha", "ouch", "yuk"),
        back_channels = c("uh[- ]huh", "uhuh", "yeah"),
        summons = "hey",
        justification = "because"
    )

    counts <- presidential_debates_2012 %>%
        with(term_count(dialogue, grouping.var = list(person, time), discoure_markers))

    counts

    ## Coverage: 100% 
    ## # A tibble: 10 x 7
    ##    person time  n.words response_cries back_channels summons justification
    ##    <fct>  <fct>   <int> <chr>          <chr>         <chr>   <chr>        
    ##  1 OBAMA  time~    3599 3(.08%)        0             43(1.1~ 26(.72%)     
    ##  2 OBAMA  time~    7477 2(.03%)        0             42(.56~ 29(.39%)     
    ##  3 OBAMA  time~    7243 1(.01%)        1(.01%)       58(.80~ 33(.46%)     
    ##  4 ROMNEY time~    4085 0              0             27(.66~ 8(.20%)      
    ##  5 ROMNEY time~    7536 1(.01%)        3(.04%)       49(.65~ 20(.27%)     
    ##  6 ROMNEY time~    8303 5(.06%)        0             84(1.0~ 19(.23%)     
    ##  7 CROWL~ time~    1672 2(.12%)        0             4(.24%) 12(.72%)     
    ##  8 LEHRER time~     765 3(.39%)        3(.39%)       0       0            
    ##  9 QUEST~ time~     583 2(.34%)        0             0       2(.34%)      
    ## 10 SCHIE~ time~    1445 0              0             2(.14%) 6(.42%)

Printing
--------

    print(counts, pretty = FALSE)

    ## Coverage: 100% 
    ## # A tibble: 10 x 7
    ##    person time  n.words response_cries back_channels summons justification
    ##    <fct>  <fct>   <int>          <int>         <int>   <int>         <int>
    ##  1 OBAMA  time~    3599              3             0      43            26
    ##  2 OBAMA  time~    7477              2             0      42            29
    ##  3 OBAMA  time~    7243              1             1      58            33
    ##  4 ROMNEY time~    4085              0             0      27             8
    ##  5 ROMNEY time~    7536              1             3      49            20
    ##  6 ROMNEY time~    8303              5             0      84            19
    ##  7 CROWL~ time~    1672              2             0       4            12
    ##  8 LEHRER time~     765              3             3       0             0
    ##  9 QUEST~ time~     583              2             0       0             2
    ## 10 SCHIE~ time~    1445              0             0       2             6

    print(counts, zero.replace = "_")

    ## Coverage: 100% 
    ## # A tibble: 10 x 7
    ##    person time  n.words response_cries back_channels summons justification
    ##    <fct>  <fct>   <int> <chr>          <chr>         <chr>   <chr>        
    ##  1 OBAMA  time~    3599 3(.08%)        _             43(1.1~ 26(.72%)     
    ##  2 OBAMA  time~    7477 2(.03%)        _             42(.56~ 29(.39%)     
    ##  3 OBAMA  time~    7243 1(.01%)        1(.01%)       58(.80~ 33(.46%)     
    ##  4 ROMNEY time~    4085 _              _             27(.66~ 8(.20%)      
    ##  5 ROMNEY time~    7536 1(.01%)        3(.04%)       49(.65~ 20(.27%)     
    ##  6 ROMNEY time~    8303 5(.06%)        _             84(1.0~ 19(.23%)     
    ##  7 CROWL~ time~    1672 2(.12%)        _             4(.24%) 12(.72%)     
    ##  8 LEHRER time~     765 3(.39%)        3(.39%)       _       _            
    ##  9 QUEST~ time~     583 2(.34%)        _             _       2(.34%)      
    ## 10 SCHIE~ time~    1445 _              _             2(.14%) 6(.42%)

Plotting
--------

    plot(counts)

![](tools/figure/unnamed-chunk-6-1.png)

    plot(counts, labels=TRUE)

![](tools/figure/unnamed-chunk-6-2.png)

    plot_ca(counts, FALSE)

![](tools/figure/unnamed-chunk-6-3.png)

Ngram Collocations
------------------

**termco** wraps the [**quanteda**](https://github.com/kbenoit/quanteda)
package to examine important ngram collocations. **quanteda**'s
`collocation` function provides measures of: `"lambda"`, `"z"`, and
`"frequency"` to examine the strength of relationship between ngrams.
**termco** adds stopword removal, min/max character filtering, and
stemming to **quanteda**'s `collocation` as well as a generic `plot`
method.

    x <- presidential_debates_2012[["dialogue"]]

    frequent_ngrams(x)

    ##            collocation length frequency count_nested    lambda         z
    ##  1:          make sure      2       127          127  7.554897 32.834995
    ##  2:    governor romney      2       105          104  9.271292 20.461487
    ##  3:         four years      2        63           63  7.338151 28.204976
    ##  4:   mister president      2        61           51  7.834853 19.748190
    ##  5:      united states      2        31           31  9.795398 17.356448
    ##  6:       middle class      2        30           30  8.777654 16.614018
    ##  7:          last four      2        27           27  6.115321 21.912251
    ##  8:    last four years      3        27            0 -1.566379 -1.028654
    ##  9:        health care      2        26           26  8.227977 20.429621
    ## 10:    american people      2        26           26  5.120440 19.048883
    ## 11:        middle east      2        26           26 10.742379  7.485044
    ## 12:   small businesses      2        22           22  7.762762 20.244536
    ## 13:        making sure      2        19           19  5.356647 17.260131
    ## 14:     million people      2        17           17  4.780434 15.493120
    ## 15: federal government      2        15           15  6.507298 17.346209
    ## 16:       young people      2        15           15  5.624489 14.208840
    ## 17:         dodd frank      2        15           15 14.718342  7.300509
    ## 18:     small business      2        13           13  7.102122 17.040580
    ## 19:      middle income      2        13           13  6.871943 15.504096
    ## 20:  governor romney's      2        13           13  8.786176  6.091802

    frequent_ngrams(x, gram.length = 3)

    ##                 collocation length frequency count_nested     lambda
    ##  1:         last four years      3        27            0 -1.5663795
    ##  2:    twenty three million      3        11            0  4.1864145
    ##  3:   thousand nine hundred      3        11            0 -0.3498540
    ##  4:   middle class families      3        10            0 -4.2845959
    ##  5:   thousand five hundred      3         8            0 -1.1540324
    ##  6:    governor romney says      3         8            0 -4.0173427
    ##  7:    three million people      3         6            0 -0.1340563
    ##  8:         next four years      3         6            0 -0.8781626
    ##  9:    governor romney said      3         6            0 -3.3551066
    ## 10:  middle income families      3         6            0 -3.9239765
    ## 11:       five million jobs      3         5            0  2.1884488
    ## 12:         five point plan      3         5            0  2.7135106
    ## 13:   seven hundred sixteen      3         5            0  0.4756980
    ## 14: hundred sixteen billion      3         5            0  0.4265593
    ## 15:    dollar seven hundred      3         5            0 -0.6058730
    ## 16:    dollar five trillion      3         5            0 -0.8280113
    ## 17:       four years closer      3         5            0 -3.0979869
    ## 18:     forty seven million      3         4            0  0.7245824
    ## 19:   best education system      3         4            0  0.2615127
    ## 20:        rising take home      3         4            0 -0.1449834
    ##               z
    ##  1: -1.02865393
    ##  2:  1.68276560
    ##  3: -0.16798792
    ##  4: -2.63216754
    ##  5: -0.70519267
    ##  6: -2.26634962
    ##  7: -0.08481962
    ##  8: -0.42487275
    ##  9: -2.02007252
    ## 10: -2.43709086
    ## 11:  1.22412938
    ## 12:  1.22017842
    ## 13:  0.21031506
    ## 14:  0.16642917
    ## 15: -0.36134012
    ## 16: -0.39457112
    ## 17: -1.44984909
    ## 18:  0.32149800
    ## 19:  0.11938982
    ## 20: -0.06723986

    frequent_ngrams(x, order.by = "lambda")

    ##                 collocation length frequency count_nested   lambda
    ##  1:              dodd frank      2        15           15 14.71834
    ##  2:         standard bearer      2         4            4 13.48186
    ##  3:           onest century      2         3            3 13.23057
    ##  4:   intellectual property      2         3            3 13.23057
    ##  5:            joint chiefs      2         3            3 13.23057
    ##  6:            apology tour      2         3            3 13.23057
    ##  7:             wall street      2         9            9 13.13031
    ##  8:          prime minister      2         2            2 12.89412
    ##  9:      permanent resident      2         2            2 12.89412
    ## 10:         abraham lincoln      2         2            2 12.89412
    ## 11:           raton florida      2         2            2 12.89412
    ## 12:         haqqani network      2         2            2 12.89412
    ## 13:              boca raton      2         2            2 12.89412
    ## 14: unintended consequences      2         2            2 12.89412
    ## 15:      appleton wisconsin      2         2            2 12.89412
    ## 16:             food stamps      2         9            9 12.61946
    ## 17:      planned parenthood      2         5            5 12.58386
    ## 18:        self deportation      2         4            4 12.38322
    ## 19:        cleveland clinic      2         3            3 12.13193
    ## 20:    religious minorities      2         3            3 12.13193
    ##            z
    ##  1: 7.300509
    ##  2: 6.561118
    ##  3: 6.390952
    ##  4: 6.390952
    ##  5: 6.390952
    ##  6: 6.390952
    ##  7: 7.886454
    ##  8: 6.147013
    ##  9: 6.147013
    ## 10: 6.147013
    ## 11: 6.147013
    ## 12: 6.147013
    ## 13: 6.147013
    ## 14: 6.147013
    ## 15: 6.147013
    ## 16: 7.972817
    ## 17: 7.455987
    ## 18: 7.285615
    ## 19: 7.060603
    ## 20: 7.060603

### Collocation Plotting

    plot(frequent_ngrams(x))

![](tools/figure/unnamed-chunk-8-1.png)

    plot(frequent_ngrams(x), drop.redundant.yaxis.text = FALSE)

![](tools/figure/unnamed-chunk-8-2.png)

    plot(frequent_ngrams(x, gram.length = 3))

![](tools/figure/unnamed-chunk-8-3.png)

    plot(frequent_ngrams(x, order.by = "lambda"))

![](tools/figure/unnamed-chunk-8-4.png)

Converting to Document Term Matrix
----------------------------------

Regular expression counts can be useful features in machine learning
models. The **tm** package's `DocumentTermMatrix` is a popular data
structure for machine learning in **R**. The `as_dtm` and `as_tdm`
functions are useful for coercing the count `data.table` structure of a
`term_count` object into a `DocumentTermMatrix`/`TermDocumentMatrix`.
The result can be combined with token/word only `DocumentTermMatrix`
structures using `cbind` & `rbind`.

    as_dtm(markers)

    ## <<DocumentTermMatrix (documents: 10, terms: 4)>>
    ## Non-/sparse entries: 27/13
    ## Sparsity           : 32%
    ## Maximal term length: 14
    ## Weighting          : term frequency (tf)

    cosine_distance <- function (x, ...) {
        x <- t(slam::as.simple_triplet_matrix(x))
        stats::as.dist(1 - slam::crossprod_simple_triplet_matrix(x)/(sqrt(slam::col_sums(x^2) %*% 
            t(slam::col_sums(x^2)))))
    }


    mod <- hclust(cosine_distance(as_dtm(markers)))
    plot(mod)
    rect.hclust(mod, k = 5, border = "red")

![](tools/figure/unnamed-chunk-9-1.png)

    (clusters <- cutree(mod, 5))

    ##     OBAMA.time 1     OBAMA.time 2     OBAMA.time 3    ROMNEY.time 1 
    ##                1                1                1                2 
    ##    ROMNEY.time 2    ROMNEY.time 3   CROWLEY.time 2    LEHRER.time 1 
    ##                2                2                3                4 
    ##  QUESTION.time 2 SCHIEFFER.time 3 
    ##                5                3

Building an Expert Rules, Regex Classifier Model
================================================

Machine learning models of classification are great when you have known
tags to train with because the model scales. Qualitative, expert based
human coding is terrific for when you have no tagged data. However, when
you have a larger, untagged data set the machine learning approaches
have no outcome to learn from and the data is too large to classify by
hand. One solution is to use a expert rules, regular expression approach
that is somewhere between machine learning and hand coding. This is one
solution for tagging larger, untagged data sets. Additionally, when each
text element contains larger chunks of text, [unsupervised clustering
type algorithms](https://github.com/trinker/clustext) such as k-means,
non-negative matrix factorization, hierarchical clustering, or [topic
modeling](https://github.com/trinker/topicmodels_learning) may be of use
for creating clusters that could be interpreted and treated as
categories.

This example section highlights the types of function combinations and
order for a typical expert rules classification. This task typically

NEWS
NEWS
====

Versioning
----------

Releases will be numbered with the following semantic versioning format:

<major>.<minor>.<patch>

And constructed with the following guidelines:

* Breaking backward compatibility bumps the major (and resets the minor
  and patch)
* New additions without breaking backward compatibility bumps the minor
  (and resets the patch)
* Bug fixes and misc changes bumps the patch



termco 0.5.0 -
----------------------------------------------------------------

BUG FIXES

* `ngram_collocations` did not properly merge the **quanteda** outputs resulting
  in the `length` column being replicated multiple times.  Additionally, length
  was integer whereas the other ngram measures are numeric resulting in a
  **data.table** warning in `melt`.  Both of these issues have been addressed.

* `colo` did not copy a single term to the clipboard with quotes.  See
  issue #50.

NEW FEATURES

* `plot_upset` added to enable exploration of overlapping instersections between
  `term_count` categories: http://caleydo.org/tools/upset.

* `get_text` added to extract the original text associated with particular tags.

* `frequent_terms_co_occurrence` added to view the co-occurrence between frequent
  terms.  A combination of `frequent_terms` and `tag_co_occurrence`.

* `term_before`, `term_after`, & `term_first` added to get frequencies of terms
  relative to other terms or specific locations.

* `token_count` added to count the occurrence of tokens within a vector of
  strings.  This function differs from`term_count` in that `term_count` is
  regex based, allowing for fuzzy matching.  This function only searches for
  lower cased tokens (words, number sequences, or punctuation) providing a well
  defined counting function that is faster than `term_count` but less flexible.

* `as_term_list` added.  This is a convenience function to convert a vector of
  terms or a quanteda `dictionary` into a named list.

* `combine_counts` added to enable combining `term_count` and `token_count`
  objects.

* `match_word` added to match words to regular expressions.  Roughly equivalent
  **qdap**'s `term_match`.

* `read_term_list`/`write_term_list` added to aid in the reading in/writing out
  and formatting of term list files.

* `classification_template` added to manually add a classification script
  template.  This template has a suggested **termco** based workflow that may be
  useful for classification projects.

* `test_regex` added to test an atomic vector, list, or term list of regexes for
  validity.

* `mutate_counts` added to apply a normalizing function to all the term columns
  of a `term_count`/`token_count` object without stripping the attributes and 
  class.

* `drop_terms` added to allow the user to explore/iterate on a term list and
  drop terms prior prior to \code{term_count} use without manually editing an
  external term list file.

* `tidy_counts` added to converts a wide matrix of counts to tidy form (tags are
  stretched long-wise with corresponding counts of tags).

* `set_meta_tags` added for setting the `metatags` attribute on a
  `term_count/`token_count` object.  This can also be controlled by separators
  in the term/token list passed to `term_count/`token_count`.
  
* `select_counts` added for safely selecting `term_count/`token_count` object
  columns without stripping attributes.  Works like `?dplyr::select`.

MINOR FEATURES

* `important_terms` picks up a plot method corresponding to the `frequent_terms`
  plot method.

* `term_count` checks for duplicate categories within tiers for hierarchical
  term lists.

* `read_term_list` checks for valid regex.

IMPROVEMENTS

* `validate_model` now uses `classify` before validating to assign tags.

* `tag_co_occurrence` used a grid + base plotting approach that required
  restarting the graphics device between plots.  This dependency has been
  replaces with a dependency on **ggraph** for plotting networks as grid
  objects.

* `plot.validate_model` now shows tag counts in the sample to provide a relative
  importance of the accuracy in making decisions.

* Open, unescaped or regexes [(i.e., `|)` unescaped pipe followed by a closing
  group character] are now caught and warned for `read_term_list` and thus
  `term_count`.
  
* `metatags` is an official attribute that can be used to group common tags 
  together.  This is common in qualitative coding where one tags text and then
  groups these subtags together into coherent metatags.  This is used by
  `tidy_counts` and can be used by other future features.

CHANGES

* The **stopwords** package replaces the **tm** package for providing default
  stopword lists.  The **stopwords** package is more comprehensive and lighter
  weight.  This changes allows the removal of the **tm** package as a dependency.
  Suggested by Ken Benoit issue #69.

* `important_terms` now uses `quanteda::dfm_tfidf` rather than `tm::weightTfIdf`.
  This means the tf-idf weighting is done is base 10 log rather than base 2 as
  done with the **tm** package.  Suggested by Ken Benoit issue #69.

* `as_dtm` & `as_tdm` moved to the **gofastr** package where they can be used by
  other packages and their classed objects.  **termco** re-exports the two
  functions.

* `summary.validate_model` used to return `n` which was the number of tags from
  the `termco` object.  It now gives n.tags and n.classified to be more explicit
  about counts of potential tags and tags actually assigned by `classify`.

* `colo` no longer uses non-standard evaluation; terms must be quoted.

* `ngram_collocations` has been renamed to  `frequent_ngrams` for better clarity
  in what the function does and as a counter part to `frequent_terms`.

* `update_names` renamed to `rename_tags` to be consistent with naming
  conventions.

* `term_cols` renamed to `tag_cols` to be consistent with naming
  conventions.
  
* `token_count` has no print method of it's own any more.  The `print` method

R/evaluate.R
#' Model Evaluation
#'
#' Get accuracy, precision, and recall for multi-class, multi-tag, predictions.
#'
#' @param x The model classification \code{\link[base]{list}}/\code{\link[base]{vector}}
#' (typically the results of \code{classify}).
#' @param known The known expert coded \code{\link[base]{list}}/\code{\link[base]{vector}} of outcomes.
#' @return Returns a list of seven elements:
#' \item{N}{The number of elements being assessed}
#' \item{confusion_matrix}{A list of confusion matrices for each tag}
#' \item{tag_accuracy}{A tag named vector of accuracies computed from the confusion matrices; (tp + tn)/(tp + tn + fp + fn)}
#' \item{tag_precision}{A tag named vector of precisions computed from the confusion matrices; tp/(tp + fp)}
#' \item{tag_recall}{A tag named vector of accuracies computed from the confusion matrices; tp/(tp + fn)}
#' \item{macro_averaged}{Macro averaged accuracy, precision, and recall; computed accuracy, precision, and recall for each confusion matrix and average}
#' \item{micro_averaged}{Micro averaged accuracy, precision, and recall; add the confusion amtrices and compute accuracy, precision, and recall}
#' @keywords accuracy model fit
#' @references https://www.youtube.com/watch?v=OwwdYHWRB5E&index=31&list=PL6397E4B26D00A269
#' @export
#' @examples
#' known <- list(1:3, 3, NA, 4:5, 2:4, 5, integer(0))
#' tagged <- list(1:3, 3, 4, 5:4, c(2, 4:3), 5, integer(0))
#' evaluate(tagged, known)
#'
#' ## Examples
#' library(dplyr)
#' data(presidential_debates_2012)
#'
#' discoure_markers <- list(
#'     response_cries = c("\\boh", "\\bah", "\\baha", "\\bouch", "yuk"),
#'     back_channels = c("uh[- ]huh", "uhuh", "yeah"),
#'     summons = "hey",
#'     justification = "because"
#' )
#'
#'
#' ## Only Single Tag Allowed Per Text Element
#' mod1 <- presidential_debates_2012 %>%
#'     with(., term_count(dialogue, TRUE, discoure_markers)) %>%
#'     classify()
#'
#' fake_known <- mod1
#' set.seed(1)
#' fake_known[sample(1:length(fake_known), 300)] <- "random noise"
#'
#' evaluate(mod1, fake_known)
#'
#' ## Multiple Tags Allowed
#' mod2 <- presidential_debates_2012 %>%
#'     with(., term_count(dialogue, TRUE, discoure_markers)) %>%
#'     classify(n = 2)
#'
#' fake_known2 <- mod2
#' set.seed(30)
#' fake_known2[sample(1:length(fake_known2), 500)] <- c("random noise", "back_channels")
#'
#' (myacc <- evaluate(mod2, fake_known2))
#' myacc$confusion_matrix
#' myacc$tag_accuracy
evaluate <- function(x, known){

    stopifnot(length(x) == length(known))

    if (!is.list(x)) x <- as.list(x)
    if (!is.list(known)) known <- as.list(known)
    x <- lapply(x, function(a) sapply(a, function(b) {b[is.na(b)] <- "No_Code_Given"; b}))
    known <- lapply(known, function(a) sapply(a, function(b) {b[is.na(b)] <- "No_Code_Given"; b}))
    out <- evaluation(x, known)
    class(out) <- "evaluate"
    out

}

#' Prints an evaluate Object
#'
#' Prints an evaluate object
#'
#' @param x The evaluate object.
#' @param digits The number of digits to print.
#' @param \ldots ignored
#' @method print evaluate
#' @export
print.evaluate <- function(x, digits = 3, ...){
#browser()
    x[['tag_level']][-1] <- lapply(x[['tag_level']][-1], digit_format, digits)
    tag_level <- sub("^\\s", "", utils::capture.output(print(x[['tag_level']], row.names=FALSE)))
    bar <- paste(rep("-", nchar(tag_level)[1]), collapse="")
    cat(bar, "\n")
    cat("Tag Level Measures\n")
    cat(bar, "\n")
    cat(paste(tag_level, collapse='\n'))
    cat('\n\n')

    n <- pn(x[["N"]])
    ner <- paste0("N:", paste(rep(" ", 15 + digits -nchar(n)), collapse = ""), n, "\n\n")
    bar2 <- paste(rep("-", nchar(ner)-2), collapse="")
    cat(bar2, "\n")
    cat("Summary Measures\n")
    cat(bar2, "\n")
    cat(ner)
    cat(        "Macro-Averaged  \n")

    filler1 <- ifelse(nchar(digit_format(x[['macro_averaged']], digits)) - 1 - digits == 0, " ", "")
    filler2 <- ifelse(nchar(digit_format(x[['micro_averaged']], digits)) - 1 - digits == 0, " ", "")
    cat(paste(
        paste0(c("  Accuracy:    ",
                 "  F-score:     ",
                 "  Precision:   ",
                 "  Recall:      "),
        filler1, digit_format(x[['macro_averaged']], digits)
    ), collapse="\n"))
    cat("\n")
    cat(        "\nMicro-Averaged  \n")
    cat(paste(
        paste0(c("  Accuracy:    ",
                 "  F-score:     ",
                 "  Precision:   ",
                 "  Recall:      "),
        filler2, digit_format(x[['micro_averaged']], digits)
    ), collapse="\n"))
    cat("\n\n")
}

confusion_matrices <- function(pred, actual){

    classes <- sort(unique(unlist(list(actual, pred))))
    cm <- matrix(rep(0, 4), ncol=2, dimnames = list(c('Pred-1', 'Pred-0'), c('Actual-1', 'Actual-0')))

    stats::setNames(lapply(classes, function(x){

        vals <- table(paste0(
            as.numeric(sapply(pred, function(y) x %in% y)),
            as.numeric(sapply(actual, function(y) x %in% y))
        ))

        try(cm[1, 1] <- vals[names(vals) == '11'], silent=TRUE)
        try(cm[1, 2] <- vals[names(vals) == '10'], silent=TRUE)
        try(cm[2, 1] <- vals[names(vals) == '01'], silent=TRUE)
        try(cm[2, 2] <- vals[names(vals) == '00'], silent=TRUE)
        cm
    }), classes)
}



evaluation <- function(pred, actual){

    out <- list()
    precision <- function(x) {out <- x[1,1]/sum(x[1,]); ifelse(is.nan(out), 0, out)}
    recall <- function(x) {out <- x[1,1]/sum(x[, 1]); ifelse(is.nan(out), 0, out)}
    F_score_fun <- function(prec, recall){out <- 2*((prec*recall)/(prec+recall));ifelse(is.nan(out), 0, out)}
    out[['N']] <- length(pred)
    cm <- out[['confusion_matrix']] <- confusion_matrices(pred, actual)

    tag_accuracy <- unlist(lapply(cm, function(x){
            sum(diag(x))/sum(x)
        }))

    tag_precision <- unlist(lapply(cm, function(x){
            precision(x)
        }))

    tag_recall <- unlist(lapply(cm, function(x){
            recall(x)
        }))

    F_score <- unlist(Map(F_score_fun, tag_precision, tag_recall))

    out[['tag_level']] <- data.frame(
        tag = names(tag_precision),
        precision = tag_precision,
        recall = tag_recall,
        F_score = F_score,
        accuracy = tag_accuracy,
        row.names = NULL
    )

    out[['macro_averaged']] <- list(
        accuracy = mean(tag_accuracy),
        F = mean(F_score),
        precision = mean(tag_precision),
        recall = mean(tag_recall)
    )

    summed_cm <- Reduce(`+`, cm)

    out[['micro_averaged']] <- list(
        accuracy = sum(diag(summed_cm))/sum(summed_cm),
        F = F_score_fun(precision(summed_cm), recall(summed_cm)),
        precision = precision(summed_cm),
        recall = recall(summed_cm)
    )

    out
}
